packages/rpc-methods/src/permitted/requestSnaps.ts

```diff
@@ -192,33 +197,27 @@ async function requestSnapsImplementation(
     } as RequestedPermissions;
     const existingPermissions = await getPermissions();
 
-    let approvedPermissions = [];
-
     if (!existingPermissions) {
-      approvedPermissions = await requestPermissions(requestedPermissions);
-    } else if (!hasRequestedSnaps(existingPermissions, requestedSnaps)) {
+      const [, metadata] = await requestPermissions(requestedPermissions);
+      res.result = metadata.data[
+        WALLET_SNAP_PERMISSION_KEY
+      ] as InstallSnapsResult;
+    } else if (hasRequestedSnaps(existingPermissions, requestedSnaps)) {
+      res.result = await handleInstallSnaps(requestedSnaps, installSnaps);
+    } else {
       const mergedPermissionsRequest = getSnapPermissionsRequest(
         existingPermissions,
         requestedPermissions,
       );
-      approvedPermissions = await requestPermissions(mergedPermissionsRequest);
-    }
 
-    if (
-      (!existingPermissions ||
-        !hasRequestedSnaps(existingPermissions, requestedSnaps)) &&
-      !approvedPermissions?.length
-    ) {
-      throw ethErrors.provider.userRejectedRequest({ data: req });
+      const [, metadata] = await requestPermissions(mergedPermissionsRequest);
```

Comment:
I'm not sure if we need all these branches 🤔

Reply:
Yes we need all of them :)